roachtest: ttl-restart/num-restart-nodes=1 failed

**Note:** This build has runtime assertions enabled. If the same failure was hit in a run without assertions enabled, there should be a similar failure without this message. If there isn&#39;t one, then this failure is likely due to an assertion violation or (assertion) timeout.

roachtest.ttl-restart/num-restart-nodes=1 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20273390?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20273390?buildTab=artifacts#/ttl-restart/num-restart-nodes=1) on master @ [8765705442d59920da3424d770a76ee50f1eee06](https://github.com/cockroachdb/cockroach/commits/8765705442d59920da3424d770a76ee50f1eee06):


```
(soon.go:60).SucceedsWithin: condition failed to evaluate within 1m0s: from ttl_restart.go:408: error waiting for TTL progress after restart: range 76 (from leaseholder 3) not found in follow-up check
(monitor.go:267).Wait: monitor failure: monitor user task failed: t.Fatal() was called
test artifacts and logs in: /artifacts/ttl-restart/num-restart-nodes=1/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>metamorphicLeases=default</code>
 - <code>runtimeAssertionsBuild=true</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20273390/ttl-restart-num-restart-nodes-1/1753946075165/1753946423025)

</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*ttl-restart/num-restart-nodes=1.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53083

roachtest.ttl-restart/num-restart-nodes=1 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20281408?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20281408?buildTab=artifacts#/ttl-restart/num-restart-nodes=1) on master @ [9831be5f895ec7758d43bbec0df76f1e14e78d44](https://github.com/cockroachdb/cockroach/commits/9831be5f895ec7758d43bbec0df76f1e14e78d44):


```
(soon.go:60).SucceedsWithin: condition failed to evaluate within 1m0s: from ttl_restart.go:408: error waiting for TTL progress after restart: range 76 (from leaseholder 3) not found in follow-up check
(monitor.go:267).Wait: monitor failure: monitor user task failed: t.Fatal() was called
test artifacts and logs in: /artifacts/ttl-restart/num-restart-nodes=1/cpu_arch=arm64/run_1
```

Parameters:
 - <code>arch=arm64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>metamorphicBufferedSender=true</code>
 - <code>metamorphicLeases=leader</code>
 - <code>metamorphicWriteBuffering=true</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20281408/ttl-restart-num-restart-nodes-1/1754031670863/1754032019490)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*ttl-restart/num-restart-nodes=1.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

roachtest.ttl-restart/num-restart-nodes=1 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAwsBazel/20287879?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAwsBazel/20287879?buildTab=artifacts#/ttl-restart/num-restart-nodes=1) on master @ [b065097dfc3353bea82cea588ce337d4a7c435c6](https://github.com/cockroachdb/cockroach/commits/b065097dfc3353bea82cea588ce337d4a7c435c6):


```
(soon.go:60).SucceedsWithin: condition failed to evaluate within 1m0s: from ttl_restart.go:408: error waiting for TTL progress after restart: range 76 (from leaseholder 3) not found in follow-up check
(monitor.go:267).Wait: monitor failure: monitor user task failed: t.Fatal() was called
test artifacts and logs in: /artifacts/ttl-restart/num-restart-nodes=1/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=aws</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>metamorphicBufferedSender=true</code>
 - <code>metamorphicLeases=default</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

_Grafana is not yet available for aws clusters_</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*ttl-restart/num-restart-nodes=1.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

**Note:** This build has runtime assertions enabled. If the same failure was hit in a run without assertions enabled, there should be a similar failure without this message. If there isn&#39;t one, then this failure is likely due to an assertion violation or (assertion) timeout.

roachtest.ttl-restart/num-restart-nodes=1 [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20289787?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20289787?buildTab=artifacts#/ttl-restart/num-restart-nodes=1) on master @ [ca0d4076c381609d6487040141c87e36a68b211f](https://github.com/cockroachdb/cockroach/commits/ca0d4076c381609d6487040141c87e36a68b211f):


```
(soon.go:60).SucceedsWithin: condition failed to evaluate within 1m0s: from ttl_restart.go:408: error waiting for TTL progress after restart: range 76 (from leaseholder 3) not found in follow-up check
(monitor.go:267).Wait: monitor failure: monitor user task failed: t.Fatal() was called
test artifacts and logs in: /artifacts/ttl-restart/num-restart-nodes=1/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>metamorphicLeases=default</code>
 - <code>runtimeAssertionsBuild=true</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)



See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-20289787/ttl-restart-num-restart-nodes-1/1754290751968/1754291098180)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*ttl-restart/num-restart-nodes=1.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/cmd/roachtest/tests/ttl_restart.go
+++ b/pkg/cmd/roachtest/tests/ttl_restart.go
@@ -53,6 +53,11 @@ const jobCheckSQL = `
   ORDER BY created DESC LIMIT 1
 `
 
+const (
+	maxRetryAttempts = 5
+	ttlJobWaitTime   = 8 * time.Minute
+)
+
 type ttlJobInfo struct {
 	JobID          int
 	CoordinatorID  int
@@ -88,112 +93,80 @@ func runTTLRestart(ctx context.Context, t test.Test, c cluster.Cluster, numResta
 		db := c.Conn(ctx, t.L(), 1)
 		defer db.Close()
 
-		t.Status("create the table")
-		setup := []string{
-			// Speed up the test by doing the replan check often and with a low threshold.
-			"SET CLUSTER SETTING sql.ttl.replan_flow_frequency = '15s'",
-			"SET CLUSTER SETTING sql.ttl.replan_flow_threshold = '0.1'",
-			// Disable the stability window to ensure immediate replanning on node changes.
-			"SET CLUSTER SETTING sql.ttl.replan_stability_window = 1",
-			// Add additional logging to help debug the test on failure.
-			"SET CLUSTER SETTING server.debug.default_vmodule = 'ttljob_processor=1,distsql_plan_bulk=1'",
-			// Create the schema to be used in the test
-			"CREATE DATABASE IF NOT EXISTS ttldb",
-			"CREATE TABLE IF NOT EXISTS ttldb.tab1 (pk INT8 NOT NULL PRIMARY KEY, ts TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP)",
-		}
-		for _, stmt := range setup {
-			if _, err := db.ExecContext(ctx, stmt); err != nil {
-				return errors.Wrapf(err, "error with statement: %s", stmt)
-			}
+		// Determine how many nodes we need TTL activity on based on restart scenario
+		requiredTTLNodes := 3 // Default for numRestartNodes=1
+		if numRestartNodes == 2 {
+			requiredTTLNodes = 2
 		}
 
-		t.Status("add manual splits so that ranges are distributed evenly across the cluster")
-		if _, err := db.ExecContext(ctx, "ALTER TABLE ttldb.tab1 SPLIT AT VALUES (6000), (12000)"); err != nil {
-			return errors.Wrapf(err, "error adding manual splits")
-		}
-
-		t.Status("insert data")
-		if _, err := db.ExecContext(ctx, "INSERT INTO ttldb.tab1 (pk) SELECT generate_series(1, 18000)"); err != nil {
-			return errors.Wrapf(err, "error ingesting data")
-		}
-
-		t.Status("relocate ranges to distribute across nodes")
-		// Moving ranges is put under a SucceedsSoon to account for errors like:
-		// "lease target replica not found in RangeDescriptor"
-		testutils.SucceedsSoon(t, func() error { return distributeLeases(ctx, t, db) })
-		leases, err := gatherLeaseDistribution(ctx, t, db)
-		if err != nil {
-			return errors.Wrapf(err, "error gathering lease distribution")
-		}
-		showLeaseDistribution(t, leases)
-
-		t.Status("enable TTL")
-		ts := db.QueryRowContext(ctx, "SELECT EXTRACT(HOUR FROM now() + INTERVAL '2 minutes') AS hour, EXTRACT(MINUTE FROM now() + INTERVAL '2 minutes') AS minute")
-		var hour, minute int
-		if err := ts.Scan(&hour, &minute); err != nil {
-			return errors.Wrapf(err, "error generating cron expression")
-		}
-		ttlCronExpression := fmt.Sprintf("%d %d * * *", minute, hour)
-		t.L().Printf("using a cron expression of '%s'", ttlCronExpression)
-		ttlJobSettingSQL := fmt.Sprintf(`ALTER TABLE ttldb.tab1
-			SET (ttl_expiration_expression = $$(ts::timestamptz + '1 minutes')$$,
-           ttl_select_batch_size=100,
-           ttl_delete_batch_size=100,
-           ttl_select_rate_limit=100,
-           ttl_job_cron='%s')`,
-			ttlCronExpression)
-		if _, err := db.ExecContext(ctx, ttlJobSettingSQL); err != nil {
-			return errors.Wrapf(err, "error setting TTL attributes")
-		}
-
-		t.Status("wait for the TTL job to start")
 		var jobInfo ttlJobInfo
-		waitForTTLJob := func() error {
-			var err error
-			jobInfo, err = findRunningJob(ctx, t, c, db, nil,
-				false /* expectJobRestart */, false /* allowJobSucceeded */)
-			return err
-		}
-		// The ttl job is scheduled to run in about 2 minutes. Wait for a little
-		// while longer to give the job system enough time to start the job in case
-		// the system is slow.
-		testutils.SucceedsWithin(t, waitForTTLJob, 8*time.Minute)
-		t.L().Printf("TTL job (ID %d) is running at node %d", jobInfo.JobID, jobInfo.CoordinatorID)
-		// Reset the connection so that we query from the coordinator as this is the
-		// only node that will stay up.
-		if err := db.Close(); err != nil {
-			return errors.Wrapf(err, "error closing connection")
-		}
-		db = c.Conn(ctx, t.L(), jobInfo.CoordinatorID)
+		var ttlNodes map[int]struct{}
 
-		t.Status("wait for TTL deletions to start happening")
-		// Take baseline once and reuse it for all progress checks
-		baseline, err := takeProgressBaseline(ctx, t, db)
-		if err != nil {
-			return errors.Wrapf(err, "error taking TTL progress baseline")
-		}
-		waitForTTLProgressAcrossAllNodes := func() error {
-			if err := checkTTLProgressAgainstBaseline(ctx, db, baseline); err != nil {
-				return errors.Wrapf(err, "error waiting for TTL progress after restart")
+		// Retry loop: attempt to setup table and get proper TTL distribution
+		for attempt := 1; attempt <= maxRetryAttempts; attempt++ {
+			if attempt > 1 {
+				t.L().Printf("Attempt %d/%d: Retrying table setup due to insufficient TTL distribution", attempt, maxRetryAttempts)
 			}
-			return nil
+
+			// Setup table and data
+			if err := setupTableAndData(ctx, t, db); err != nil {
+				return err
+			}
+
+			// Enable TTL and wait for job to start
+			var err error
+			jobInfo, err = enableTTLAndWaitForJob(ctx, t, c, db)
+			if err != nil {
+				return err
+			}
+
+			// Reset the connection so that we query from the coordinator as this is the
+			// only node that will stay up.
+			if err := db.Close(); err != nil {
+				return errors.Wrapf(err, "error closing connection")
+			}
+			db = c.Conn(ctx, t.L(), jobInfo.CoordinatorID)
+
+			t.Status("check TTL activity distribution across nodes")
+			err = testutils.SucceedsWithinError(func() error {
+				var err error
+				ttlNodes, err = findNodesWithJobLogs(ctx, t, c, jobInfo.JobID)
+				if err != nil {
+					return err
+				}
+				if len(ttlNodes) < requiredTTLNodes {
+					return errors.Newf("TTL activity found on only %d nodes (need %d)", len(ttlNodes), requiredTTLNodes)
+				}
+				return nil
+			}, 1*time.Minute)
+
+			if err == nil {
+				// Success! TTL is distributed across enough nodes
+				t.L().Printf("TTL job %d found on nodes: %v", jobInfo.JobID, ttlNodes)
+				break
+			}
+
+			// Handle the error
+			if ttlNodes != nil && len(ttlNodes) < requiredTTLNodes {
+				t.L().Printf("Attempt %d/%d: TTL job %d found on nodes: %v", attempt, maxRetryAttempts, jobInfo.JobID, ttlNodes)
+				t.L().Printf("Attempt %d/%d: TTL activity found on only %d nodes (need %d)", attempt, maxRetryAttempts, len(ttlNodes), requiredTTLNodes)
+
+				if attempt == maxRetryAttempts {
+					// Final attempt failed - exit successfully as current behavior
+					t.L().Printf("After %d attempts, TTL activity found on only %d nodes (need %d for restart test). Test completed successfully.", maxRetryAttempts, len(ttlNodes), requiredTTLNodes)
+					return nil
+				}
+				// Continue to next attempt
+				continue
+			}
+
+			// Other error that's not related to TTL distribution
+			return errors.Wrapf(err, "error waiting for TTL activity distribution")
 		}
-		testutils.SucceedsWithin(t, waitForTTLProgressAcrossAllNodes, 1*time.Minute)
 
 		t.Status("stop non-coordinator nodes")
 		nonCoordinatorCount := c.Spec().NodeCount - 1
 		stoppingAllNonCoordinators := numRestartNodes == nonCoordinatorCount
-		var ttlNodes map[int]struct{}
-		if !stoppingAllNonCoordinators {
-			// We need to stop a node that actually executed part of the TTL job.
-			// Relying on SQL isn't fully reliable due to potential cache staleness.
-			// Instead, we scan cockroach.log files for known TTL job log markers to
-			// identify nodes that were truly involved in the job execution.
-			ttlNodes, err = findNodesWithJobLogs(ctx, t, c, jobInfo.JobID)
-			if err != nil {
-				return errors.Wrapf(err, "error finding nodes with job logs")
-			}
-		}
 		stoppedNodes := make([]int, 0)
 		for node := 1; node <= c.Spec().NodeCount && len(stoppedNodes) < numRestartNodes; node++ {
 			if node == jobInfo.CoordinatorID {
@@ -214,6 +187,7 @@ func runTTLRestart(ctx context.Context, t test.Test, c cluster.Cluster, numResta
 
 		// If we haven't lost quorum, then the TTL job should restart and continue
 		// working before restarting the down nodes.
+		var err error
 		if numRestartNodes <= c.Spec().NodeCount/2 {
 			t.Status("ensure TTL job restarts")
 			testutils.SucceedsWithin(t, func() error {
@@ -315,108 +289,6 @@ func distributeLeases(ctx context.Context, t test.Test, db *gosql.DB) error {
 	}
 	return nil
 
-}
-
-// takeProgressBaseline captures the initial key counts for each range and its leaseholder.
-// This baseline will be used later to check if TTL progress is being made.
-func takeProgressBaseline(
-	ctx context.Context, t test.Test, db *gosql.DB,
-) (map[int]map[int]int, error) {
-	query := `
-		WITH r AS (
-			SHOW RANGES FROM TABLE ttldb.tab1 WITH DETAILS
-		)
-		SELECT
-		  range_id,
-			lease_holder,
-			count(*) AS key_count
-		FROM
-			r,
-			LATERAL crdb_internal.list_sql_keys_in_range(range_id)
-		GROUP BY
-		  range_id,
-			lease_holder
-		ORDER BY
-		  range_id`
-
-	// Map of leaseholder -> rangeID -> keyCount
-	baseline := make(map[int]map[int]int)
-
-	rows, err := db.QueryContext(ctx, query)
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-
-	for rows.Next() {
-		var rangeID, leaseHolder, keyCount int
-		if err := rows.Scan(&rangeID, &leaseHolder, &keyCount); err != nil {
-			return nil, err
-		}
-		if _, ok := baseline[leaseHolder]; !ok {
-			baseline[leaseHolder] = make(map[int]int)
-		}
-		baseline[leaseHolder][rangeID] = keyCount
-	}
-
-	return baseline, nil
-}
-
-// checkTTLProgressAgainstBaseline checks if each leaseholder has made progress
-// on at least one of their original ranges compared to the provided baseline.
-func checkTTLProgressAgainstBaseline(
-	ctx context.Context, db *gosql.DB, baseline map[int]map[int]int,
-) error {
-	query := `
-		WITH r AS (
-			SHOW RANGES FROM TABLE ttldb.tab1 WITH DETAILS
-		)
-		SELECT
-		  range_id,
-			lease_holder,
-			count(*) AS key_count
-		FROM
-			r,
-			LATERAL crdb_internal.list_sql_keys_in_range(range_id)
-		GROUP BY
-		  range_id,
-			lease_holder
-		ORDER BY
-		  range_id`
-
-	current := make(map[int]int) // rangeID -> keyCount
-
-	rows, err := db.QueryContext(ctx, query)
-	if err != nil {
-		return err
-	}
-	defer rows.Close()
-
-	for rows.Next() {
-		var rangeID, leaseHolder, keyCount int
-		if err := rows.Scan(&rangeID, &leaseHolder, &keyCount); err != nil {
-			return err
-		}
-		current[rangeID] = keyCount
-	}
-
-	for leaseHolder, ranges := range baseline {
-		madeProgress := false
-		for rangeID, oldCount := range ranges {
-			newCount, ok := current[rangeID]
-			if !ok {
-				return errors.Newf("range %d (from leaseholder %d) not found in follow-up check", rangeID, leaseHolder)
-			}
-			if newCount < oldCount {
-				madeProgress = true
-			}
-		}
-		if !madeProgress {
-			return errors.Newf("leaseholder %d made no progress on any of their original ranges", leaseHolder)
-		}
-	}
-
-	return nil
 }
 
 // findRunningJob checks the current state of the TTL job and returns metadata
@@ -526,7 +398,96 @@ func findNodesWithJobLogs(
 		nodeList = append(nodeList, node)
 	}
 	sort.Ints(nodeList)
-	t.L().Printf("TTL job %d found on nodes: %v", jobID, nodeList)
 
 	return nodesWithJob, nil
+}
+
+// setupTableAndData creates the ttldb database and tab1 table, inserts data,
+// and distributes ranges across nodes. This function can be called multiple
+// times to recreate the table for retry scenarios.
+func setupTableAndData(ctx context.Context, t test.Test, db *gosql.DB) error {
+	t.Status("create/recreate the table")
+	setup := []string{
+		// Speed up the test by doing the replan check often and with a low threshold.
+		"SET CLUSTER SETTING sql.ttl.replan_flow_frequency = '15s'",
+		"SET CLUSTER SETTING sql.ttl.replan_flow_threshold = '0.1'",
+		// Disable the stability window to ensure immediate replanning on node changes.
+		"SET CLUSTER SETTING sql.ttl.replan_stability_window = 1",
+		// Add additional logging to help debug the test on failure.
+		"SET CLUSTER SETTING server.debug.default_vmodule = 'ttljob_processor=1,distsql_plan_bulk=1'",
+		// Drop existing table if it exists
+		"DROP TABLE IF EXISTS ttldb.tab1",
+		// Create the schema to be used in the test
+		"CREATE DATABASE IF NOT EXISTS ttldb",
+		"CREATE TABLE IF NOT EXISTS ttldb.tab1 (pk INT8 NOT NULL PRIMARY KEY, ts TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP)",
+	}
+	for _, stmt := range setup {
+		if _, err := db.ExecContext(ctx, stmt); err != nil {
+			return errors.Wrapf(err, "error with statement: %s", stmt)
+		}
+	}
+
+	t.Status("add manual splits so that ranges are distributed evenly across the cluster")
+	if _, err := db.ExecContext(ctx, "ALTER TABLE ttldb.tab1 SPLIT AT VALUES (6000), (12000)"); err != nil {
+		return errors.Wrapf(err, "error adding manual splits")
+	}
+
+	t.Status("insert data")
+	if _, err := db.ExecContext(ctx, "INSERT INTO ttldb.tab1 (pk) SELECT generate_series(1, 18000)"); err != nil {
+		return errors.Wrapf(err, "error ingesting data")
+	}
+
+	t.Status("relocate ranges to distribute across nodes")
+	// Moving ranges is put under a SucceedsSoon to account for errors like:
+	// "lease target replica not found in RangeDescriptor"
+	testutils.SucceedsSoon(t, func() error { return distributeLeases(ctx, t, db) })
+	leases, err := gatherLeaseDistribution(ctx, t, db)
+	if err != nil {
+		return errors.Wrapf(err, "error gathering lease distribution")
+	}
+	showLeaseDistribution(t, leases)
+
+	return nil
+}
+
+// enableTTLAndWaitForJob enables TTL on the table with a cron expression
+// set to run in about 2 minutes, then waits for the job to start.
+func enableTTLAndWaitForJob(
+	ctx context.Context, t test.Test, c cluster.Cluster, db *gosql.DB,
+) (ttlJobInfo, error) {
+	var jobInfo ttlJobInfo
+
+	t.Status("enable TTL")
+	ts := db.QueryRowContext(ctx, "SELECT EXTRACT(HOUR FROM now() + INTERVAL '2 minutes') AS hour, EXTRACT(MINUTE FROM now() + INTERVAL '2 minutes') AS minute")
+	var hour, minute int
+	if err := ts.Scan(&hour, &minute); err != nil {
+		return jobInfo, errors.Wrapf(err, "error generating cron expression")
+	}
+	ttlCronExpression := fmt.Sprintf("%d %d * * *", minute, hour)
+	t.L().Printf("using a cron expression of '%s'", ttlCronExpression)
+	ttlJobSettingSQL := fmt.Sprintf(`ALTER TABLE ttldb.tab1
+		SET (ttl_expiration_expression = $$(ts::timestamptz + '1 minutes')$$,
+          ttl_select_batch_size=100,
+          ttl_delete_batch_size=100,
+          ttl_select_rate_limit=100,
+          ttl_job_cron='%s')`,
+		ttlCronExpression)
+	if _, err := db.ExecContext(ctx, ttlJobSettingSQL); err != nil {
+		return jobInfo, errors.Wrapf(err, "error setting TTL attributes")
+	}
+
+	t.Status("wait for the TTL job to start")
+	waitForTTLJob := func() error {
+		var err error
+		jobInfo, err = findRunningJob(ctx, t, c, db, nil,
+			false /* expectJobRestart */, false /* allowJobSucceeded */)
+		return err
+	}
+	// The ttl job is scheduled to run in about 2 minutes. Wait for a little
+	// while longer to give the job system enough time to start the job in case
+	// the system is slow.
+	testutils.SucceedsWithin(t, waitForTTLJob, ttlJobWaitTime)
+	t.L().Printf("TTL job (ID %d) is running at node %d", jobInfo.JobID, jobInfo.CoordinatorID)
+
+	return jobInfo, nil
 }

